### PR TITLE
Mrc 2201 - Push outputs to ADR: front end UI

### PR DESCRIPTION
**Pull request by carbowizzy**
## Description

This PR adds frontend for uploading outputs to ADR server. This ticket requires https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2232 ticket to implement the submit action. 

## Type of version change
_Delete as appropriate_

Minor

## Checklist

- [x] I have incremented version number, or version needs no increment
- [x] The build passed successfully, or failed because of ADR tests

### CODE REVIEW

**Review by EmmaLRussell — CHANGES_REQUESTED** [review]
Dialog looks great. Could we make all upload files selected by default?

**Review by mwoodbri — COMMENTED** [review]
Just a couple of tweaks to follow Emma's comments

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
+
+
 # hint 1.20.0
 
+* add upload button and dialog box for result outputs to ADR# hint 1.20.1
```
Comment (mwoodbri) [inline NEWS.md:5]: Something's gone a bit wrong here!

File: src/app/static/src/app/types.ts [hunk 2]
```diff
@@ -134,3 +134,13 @@ export interface ADRSchemas {
     outputZip: string
     outputSummary: string
 }
+export type ADRUploadMetadataDict = Dict<ADRUploadMetadata>;
+
+export interface ADRUploadMetadata {
```
[the remark this answers is not in the document]
Comment (mwoodbri) [reply src/app/static/src/app/types.ts:139]: I've also used `UploadFile` in mrc-2233

File: src/app/static/src/tests/components/downloadResults/downloadResults.test.ts [hunk 3]
```diff
@@ -35,14 +36,39 @@ describe("Download Results component", () => {
...
         expectTranslated(links.at(0), "Export", "Exporter", store);
         expect(links.at(0).attributes().href).toEqual("/download/spectrum/testId");
         expectTranslated(links.at(1), "Download", "Télécharger", store);
         expect(links.at(1).attributes().href).toEqual("/download/coarse-output/testId");
         expectTranslated(links.at(2), "Download", "Télécharger", store);
         expect(links.at(2).attributes().href).toEqual("/download/summary/testId")
+        expectTranslated(links.at(3), "Upload", "Télécharger", store);
     });
+
+    it(`renders, opens and closes dialog as expected`, async() => {
```
[the remark this answers is not in the document]
Comment (mwoodbri) [reply src/app/static/src/tests/components/downloadResults/downloadResults.test.ts:53]: And if you fix this then Codecov will be fixed too!

**Review by mwoodbri — CHANGES_REQUESTED** [review]
Just a few changes to make - please be sure to check for consistency with mrc-2233 (#477) where a couple of the modifications have been implemented in order to integrate with Emma's code

File: src/app/static/src/app/components/downloadResults/UploadModal.vue [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                outputFileMetadata: fakeMetadata
+            }
+        },
+        methods: {
+            handleCancel() {
+                this.$emit("close")
+            },
+            getTranslatedLabel(name: string) {
+                console.log(this.pushFilesToAdr.length)
+                return `outputFile${name.trim().replace(/\s/g, "")}`
```
[the remark this answers is not in the document]
Comment (carbowizzy) [reply src/app/static/src/app/components/downloadResults/UploadModal.vue:118]: some labels are two words. I may not need to append outputFile but the given words need to be joined.

**Comment by carbowizzy** [comment]
> Dialog looks great. Could we make all upload files selected by default?


I tried to fix all the known issue but if anything was missed, please let me know. Thanks.

File: src/app/static/src/app/components/downloadResults/UploadModal.vue [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            <div class="container">
+                <div id="dataset-id" class="mt-4">
+                    <span v-translate="'uploadFileDataset'"></span>
+                    <span>{{ dataset }}</span></div>
+                <div id="instructions" class="mt-3" v-translate="'uploadFileInstruction'"></div>
+                <div id="output-file-id" class="mt-3" v-for="uploadFile in uploadFiles" :key="uploadFile.index">
+                    <div class="mt-3 form-check">
+                        <input class="form-check-input"
+                               type="checkbox"
+                               :value="uploadFile.index"
```
Comment (mwoodbri) [inline src/app/static/src/app/components/downloadResults/UploadModal.vue:14]: This should be the key of the map entry (e.g. `outputSummary`), not the index of the value - see comment on test below

File: src/app/static/src/tests/components/downloadResults/uploadModal.test.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        const wrapper = getWrapper()
+        const store = wrapper.vm.$store
+
+        const inputs = wrapper.findAll("input.form-check-input")
+        expect(inputs.length).toBe(2)
+        inputs.at(0).setChecked(true)
+        inputs.at(1).setChecked(true)
+
+        expect(wrapper.find("#dialog").exists()).toBe(true)
+        expect(wrapper.vm.$data.uploadFilesToAdr).toMatchObject([1, 2])
```
Comment (mwoodbri) [inline src/app/static/src/tests/components/downloadResults/uploadModal.test.ts:143]:
```suggestion
        expect(wrapper.vm.$data.uploadFilesToAdr).toMatchObject(["outputZip", ""outputSummary"])
```

File: src/app/static/src/tests/components/downloadResults/uploadModal.test.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        const wrapper = getWrapper()
+        const store = wrapper.vm.$store
+
+        const overwriteText = wrapper.findAll("small")
+        expect(overwriteText.length).toBe(1)
+        expect(overwriteText.at(0).text()).toBe("This file already exists on ADR " +
+            "and will be overwritten. File was updated 25/01/2021 06:34:12")
+
+        const inputs = wrapper.findAll("input.form-check-input")
+        expect(inputs.length).toBe(2)
```
Comment (mwoodbri) [inline src/app/static/src/tests/components/downloadResults/uploadModal.test.ts:124]: You should also check that the checkboxes are ticked (the default behaviour)

File: src/app/static/src/app/components/downloadResults/UploadModal.vue [hunk 8]
```diff
@@ -0,0 +1,112 @@
+<template>
+    <div id="dialog">
+        <modal :open="open">
+            <h4 v-translate="'uploadFileToAdr'"></h4>
+            <div class="container">
+                <div id="dataset-id" class="mt-4">
+                    <span v-translate="'uploadFileDataset'"></span>
+                    <span>{{ dataset }}</span></div>
+                <div id="instructions" class="mt-3" v-translate="'uploadFileInstruction'"></div>
+                <div id="output-file-id" class="mt-3" v-for="uploadFile in uploadFiles" :key="uploadFile.index">
```
Comment (mwoodbri) [inline src/app/static/src/app/components/downloadResults/UploadModal.vue:10]: I think you should be iterating in order of the `index` field of the values in `uploadFiles` - it that right @EmmaLRussell?

File: src/app/static/src/app/components/downloadResults/UploadModal.vue [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    export default Vue.extend<Data, Methods, Computed, Props>({
+        name: "UploadModal",
+        props: {
+            open: {
+                type: Boolean
+            }
+        },
+        data(): Data {
+            return {
+                uploadFilesToAdr: [],
```
Comment (mwoodbri) [inline src/app/static/src/app/components/downloadResults/UploadModal.vue:87]: As mentioned above, I think this should be:
```suggestion
                uploadFilesToAdr: string[],
```

File: src/app/static/src/app/components/downloadResults/UploadModal.vue [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            lastModified: function (date: string) {
+                return formatDateTime(date)
+            }
+        },
+        computed: {
+            ...mapStateProps<BaselineState, keyof Computed>("baseline", {
+                dataset: state => state.selectedDataset?.title
+            }),
+            uploadFiles: mapStateProp<ADRState, Dict<UploadFile>>("adr",
+                (state: ADRState) => state.uploadFiles!
```
Comment (mwoodbri) [inline src/app/static/src/app/components/downloadResults/UploadModal.vue:104]: My PR will ensure this is populated this by triggering the relevant action on mount

**Review by mwoodbri — CHANGES_REQUESTED** [review]
Nearly there - just a couple of deviations from design regarding contents of `Data`, ordering of fields, and automatically ticking checkboxes

File: src/app/static/src/app/components/downloadResults/UploadModal.vue (hunk 10 above)
Comment (carbowizzy) [reply src/app/static/src/app/components/downloadResults/UploadModal.vue:104]: This is already being populated, mounted in download page. Emma and I discussed adding it to my ticket, sorry for not informing you.

File: src/app/static/src/app/components/downloadResults/UploadModal.vue (hunk 8 above)
Comment (carbowizzy) [reply src/app/static/src/app/components/downloadResults/UploadModal.vue:10]: I don't mind changing this to the display name but I believe Emma wanted it to be index.

File: src/app/static/src/tests/components/downloadResults/uploadModal.test.ts (hunk 7 above)
Comment (carbowizzy) [reply src/app/static/src/tests/components/downloadResults/uploadModal.test.ts:124]: Oh yeah, missed this.

File: src/app/static/src/tests/components/downloadResults/uploadModal.test.ts (hunk 6 above)
Comment (carbowizzy) [reply src/app/static/src/tests/components/downloadResults/uploadModal.test.ts:143]: I don't mind changing this to the display name but I believe Emma wanted it to be index.

File: src/app/static/src/tests/components/downloadResults/downloadResults.test.ts [hunk 11]
```diff
@@ -41,21 +44,58 @@ describe("Download Results component", () => {
...
+
+        expect(wrapper.find(UploadModal).exists()).toBe(true)
+        expect(wrapper.vm.$data.uploadModalOpen).toBe(false)
+
+        const upload = wrapper.find("#upload").find("a")
+        await upload.trigger("click")
+        expect(wrapper.vm.$data.uploadModalOpen).toBe(true)
+
+        const modal = wrapper.find(UploadModal)
+        await modal.vm.$emit("close")
```
[the remark this answers is not in the document]
Comment (carbowizzy) [reply src/app/static/src/tests/components/downloadResults/downloadResults.test.ts:83]: This is impossible with the current implementation as the modal is a child component. The only way is to stub the entire component which I doubt is worth the effort considering the fact that it's been tested differently already.

File: src/app/static/src/tests/components/downloadResults/uploadModal.test.ts (hunk 6 above)
Comment (EmmaLRussell) [reply src/app/static/src/tests/components/downloadResults/uploadModal.test.ts:143]: Sorry for confusion, it should be the keys from the dictionary (not the display name in the object) which gets set as uploadFilesToAdr. The indexes should just be used to order the files in the dialog

File: src/app/static/src/app/components/downloadResults/UploadModal.vue (hunk 10 above)
Comment (EmmaLRussell) [reply src/app/static/src/app/components/downloadResults/UploadModal.vue:104]: Sorry @mwoodbri I'd forgotten you'd picked up that ticket. Lekan and I were discussing how to modify his previous fake upload files data to conform with current master, and it seemed that might be easier to just use the state values directly.

**Review by mwoodbri — CHANGES_REQUESTED** [review]
Code looks good but I'm having some trouble testing this - not sure why! By doing some console logging I can see that `hasUploadPermission` is returning false when component is first mounted, then `adr/actions/getUserCanUpload` sets `canUpload` as expected, but the upload button doesn't then appear as expected. In other words, the button doesn't seem to react to change in global state. Any ideas?

Version and NEWS need updating, just to reflect changes on master.

**Comment by carbowizzy** [comment]
> Code looks good but I'm having some trouble testing this - not sure why! By doing some console logging I can see that `hasUploadPermission` is returning false when component is first mounted, then `adr/actions/getUserCanUpload` sets `canUpload` as expected, but the upload button doesn't then appear as expected. In other words, the button doesn't seem to react to change in global state. Any ideas?
> 
> Version and NEWS need updating, just to reflect changes on master.

I see what you mean, getUserCanUpload action takes a second to set the permission. This is the cause of the reaction. We can think about using disable and enable to control the behaviour.

**Comment by mwoodbri** [comment]
> I see what you mean, getUserCanUpload action takes a second to set the permission. This is the cause of the reaction. We can think about using disable and enable to control the behaviour but that's another discussion.

No major problem with button appearing after a brief delay (we could consider a spinner for this in the future) but on my machine the button doesn't appear at all. Is that what I should be expecting? I'm using the Antarctica dataset.

**Review by EmmaLRussell — APPROVED** [review]
Works for me.. :)

**Review by mwoodbri — APPROVED** [review]
Mystery solved - my debugging statement for ADR was:
```javascript
console.log("getUserCanUpload", true)
```
How embarrassing.

This can be merged IMO. And in the meantime I'll get a proper ADR account...

**Pull request merged**